src/sqlfluff/parser/segments_base.py

```diff
@@ -15,8 +15,11 @@
 import logging
 
 from io import StringIO
+from typing import Optional
+
 from benchit import BenchIt
 
+from .grammar import BaseGrammar

```

Comment:
I believe this is causing the [CI failure](https://app.circleci.com/pipelines/github/sqlfluff/sqlfluff/1055/workflows/b6345748-ded4-408d-9836-01a46989c4e7/jobs/7417), through a circular import? The `grammer.py` module also imports from the `segments_base.py` module: https://github.com/sqlfluff/sqlfluff/blob/master/src/sqlfluff/parser/grammar.py#L5

Reply:
I was trying to fix this last night by using [forward references](https://www.python.org/dev/peps/pep-0484/#forward-references) but so far had no luck :( 

I want to avoid conditional imports if we can (i.e. checking [typing.TYPE_CHECKING](https://docs.python.org/3/library/typing.html#typing.TYPE_CHECKING) variable before importing)

Reply:
Update: Still no luck with forward references. It's looking like eliminating this circular dependency without doing a conditional import would require considerable re-design which is not a massive priority. 

Do you have any opinions on using `typing.TYPE_CHECKING` variable?

Reply:
Yeah, you've found the tricky bit here @dnshio . [btw, welcome to the project 🙏 ]

The problem here is that *classes* which inherit from `BaseSegment` and *instances of classes* which inherit from `BaseGrammar` share a common signature. I'd love to have a common ancestor of both to standardize that interface, but my python-fu isn't good enough to come up with a smart way of setting up the inheritance so that works. As an example you'll notice that in `BaseGrammar.match()` is an instance method, while `BaseSegment.match()` is a classmethod.

Ideally that common ancestor (or mixin?) should define this common signature and live in a separate file to both `BaseSegment` and `BaseGrammar`, where when type checking we check that the relevant attributes are of a type compatible with that common ancestor.

@dnshio - any ideas on the best way to achieve that?

Reply:
> Looks like python 3.5 has an issue with that syntax, it must not have been added until 3.6?
Looks like type hints for variables were added in version 3.6. https://www.python.org/dev/peps/pep-0526/

On the other hand, Python 3.5 has reached end of life 5 days ago 🤷‍♂️  : https://www.python.org/downloads/release/python-3510/ 

We are actually forced to make a decision here. We either have to drop python 3.5 or avoid type hinting variables as you suggested in your comment. 

Unfortunately we cannot avoid type hinting some of the class vars in `BaseSegment` because we are declaring the class vars in with a default `None` but later in inheriting classes like `ArrayAccessorSegment` assign a value that is not `None`. mypy complains about that. 

```
src/sqlfluff/dialects/dialect_ansi.py:293: error: Incompatible types in assignment (expression has type "Bracketed", base class "BaseSegment" defined the type as "None")
```

In python 3.6+ we can avoid the circular import setting the type hint to `Optional[Any] = None`.

Reply:
Though this certainly deserves it's own discussion, I could get behind the idea of only supporting _current_ versions of python (i.e dropping support for `3.5`). This would allow us to use f-strings and typing. 
I also understand the value in not pulling the rug out from underneath your users in terms of what versions you support

Reply:
For what it's worth, dbt dropped support for python 3.5 some time ago. I'd get behind dropping it to properly support typing.

Reply:
@alanmcruickshank I'm still trying to wrap my head around the intention there. `BaseSegment` and `BaseGrammar` are two distinct bases that happen to share some common traits. A mixin could have been a good choice if we wanted to enforce the same interface for two classes. However, the fact that one is a classmethod and the other is an instance method means that they are not the same signature at all. They just happen to share _similar_ arguments (`cls != self` as you know). 

Having said that, mypy might be a step in the right direction towards what we want to do. Without in-depth understanding, it's hard to suggest what's the best approach. But I'd be happy to have a stab at it in another PR. 

As outlined in my previous comment, there are some workarounds for this circular import issue for now. The major blocker is Python 3.5

Reply:
I'm happy dropping support for 3.5, but we should do it as a major release if we do.

Reply:
I think it would make sense to do as part of this PR too so the reason is associated

Reply:
I could see dropping support in this PR. If you prefer @alanmcruickshank I can submit a separate PR for dropping 3.5 support. I'm guessing that entails:
- Modifying the `setup.py` `python_requires`
- Modifying the `tox.ini` by removing the python 3.5 command. 

I could also see adding the proper 3.9 version as well (as mentioned in #461)

Your call ☎️ 😁

Reply:
Lets drop 3.5 and upgrade the 3.9 version at the same time.

Looking at: https://pypistats.org/packages/sqlfluff it looks like 3.5 usage is pretty low. 

can you add a similar warning to the one in `sqlfluff.__init__` so that users get a sensible error message if they are running on 3.5 that support was dropped in `0.4.0`, so they should use a version before that if they want to use sqlfluff.

Other than that we should also:
- update `.circleci/config.yml` to reflect the drop of 3.5.
- make a note on the `README.md` that support for 3.5 will be dropped in 0.4.0

Reply:
Just to clarify, would you like this tackled in a separate PR?